// shopify-admin-api/dist/esm/services/discounts.js
import { BaseService } from '../infrastructure';
/**
 * A service for manipulating Shopify discounts.
 */
export class Discounts extends BaseService {
    constructor(shopDomain, accessToken) {
        super(shopDomain, accessToken, "discounts");
    }
    /**
      * Creates a new discount.
      */
    create(discount) {
        return this.createRequest("POST", ".json", "discount", { discount: discount });
    }
    /**
     * Gets a list of up to 250 of the shop's discounts.
     * @param options Options for filtering the results.
     */
    list(options) {
        return this.createRequest("GET", ".json", "discounts", options);
    }
    /**
     * Retrieves the discount with the given id.
     * @param options Options for filtering the results.
     */
    get(id) {
        return this.createRequest("GET", `${id}.json`, "discount");
    }
    /**
     * Enables a discount.
     */
    enable(id) {
        return this.createRequest("POST", `${id}/enable.json`, "discount");
    }
    /**
     * Disable a discount.
     */
    disable(id) {
        return this.createRequest("POST", `${id}/disable.json`, "discount");
    }
    /**
     * Deletes the discount with the given id.
     */
    delete(id) {
        return this.createRequest("DELETE", `${id}.json`);
    }
}
export default Discounts;
